promql/engine.go

```diff
@@ -1341,6 +1364,7 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 				enh.Ts = ts
 				// Make the function call.
 				outVec := call(inArgs, e.Args, enh)
+				ev.samplesStats.IncrementSamples(ts, len(points))
```

Comment:
I am curious to know the overhead of calculating per step stats for every query. Calculating it only if detailed stat is asked could be a good idea.

Reply:
I tried to rebase this CR and it got closed, i will create a new one.

I agree with the detailed but, seems that to do this i will have to change the [NewInstantQuery](https://github.com/prometheus/prometheus/blob/579331446a04503b2b1515767e7ceaceab2c6342/web/api/v1/api.go#L380) and [NewRangeQuery](https://github.com/prometheus/prometheus/blob/579331446a04503b2b1515767e7ceaceab2c6342/web/api/v1/api.go#L455) methods as the stats are always calculated  and only "stripped from the response": 

https://github.com/prometheus/prometheus/blob/579331446a04503b2b1515767e7ceaceab2c6342/web/api/v1/api.go#L477-L479

Another possibility is to some field on the [qry](https://github.com/prometheus/prometheus/blob/579331446a04503b2b1515767e7ceaceab2c6342/web/api/v1/api.go#L380) after the creation with this information... what u think?

Reply:
If the overhead is small, we can set to always calculate it. It should be fine to change the function signature of instant query and range query too. Maybe introduce a new query opts structs so that we dont have to change the signature further in the future for more options.

> Another possibility is to some field on the [qry](https://github.com/prometheus/prometheus/blob/579331446a04503b2b1515767e7ceaceab2c6342/web/api/v1/api.go#L380) after the creation with this information... what u think?

I don't think I understood what you meant here :/

Reply:
Something like.. 

```
qry, err := api.QueryEngine.NewInstantQuery(api.Queryable, r.FormValue("query"), ts)
qry.WithStats("detailed")
```

Is not ideal.. but we would not break the interface i guess.

Reply:
We have followed of passing options to the method everywhere. But this should be fine as well. But if we have more options in future, WithStats() will have to go away.